README improvements (#245)
camunda-cloud/zeebe-process-test: README.md

> [![Maven Central](https://img.shields.io/maven-central/v/io.camunda/zeebe-process-test-root)](https://search.maven.org/search?q=g:io.camunda%20zeebe-process-test)

Reviewer: ```suggestion
[![Maven Central](https://img.shields.io/maven-central/v/io.camunda/zeebe-process-test-root)](https://search.maven.org/search?q=g:io.camunda%20zeebe-process-test-root)
```

Author: I did this on purpose. The difference is the list of dependencies you will see when you click on the badge. If you put `zeebe-process-test-root` here, you will only see this dependency. By setting it to `zeebe-process-test` it will give a list of all `zeebe-process-test*` dependencies that are in the `io.camunda` group.

### Dependency

> #### Testcontainers (JDK 8+)

Reviewer: :wrench: Add an introductory sentence under _Dependency_ that the dependencies depend on the Java version

> If you are building your project with a JDK that's lower than 17 you should use this dependency. It

Reviewer: ```suggestion
If you are building your project with a JDK that's lower than 17 you need to use this dependency. It
```

starts a testcontainer in which a Zeebe test engine is running. The advantage of using this version

> instead of the embedded version is that you will be completely separated from the Java version that is

Reviewer: ```suggestion
instead of the embedded version is that your code can be implemented independently from the Java version that is
```

> used by the Zeebe engine. This does come at a cost. Testcontainers provide some overhead, which means

Reviewer: ```suggestion
used by the Zeebe engine. This has some downsides: Testcontainers provide some overhead, which means
```

tests will be slower. There is also the extra requirement that Docker must be running to execute the tests.

```xml
<dependency>
  <groupId>io.camunda</groupId>
  <artifactId>zeebe-process-test-extension-testcontainer</artifactId>
  <version>X.Y.Z</version>
  <scope>test</scope>
</dependency>
```

#### Embedded (JDK 17+)

If you are building your project with JDK 17+ you can make use of an embedded Zeebe test engine. The
advantage of using this instead of the testcontainer version is that this is the faster solution.
This also does not require Docker to be running. There is also a downside to this solution. The JDK

> requirement is bound to the Java version of the Zeebe engine. Whenever this Java version gets updated,

Reviewer: ```suggestion
requirement is bound to the Java version of the Zeebe engine. Whenever this Java version changes,
```

you'd either have to [switch to the testcontainer version](#switching-between-testcontainers-and-embedded),

> or upgrade your own JDK to match the Zeebe engine.

Reviewer: ```suggestion
or upgrade your own JDK to match Zeebe engine.
```

```xml
<dependency>
  <groupId>io.camunda</groupId>
  <artifactId>zeebe-process-test-extension</artifactId>
  <version>X.Y.Z</version>
  <scope>test</scope>
</dependency>
```

### Annotation

Annotate your test class with the `@ZeebeProcessTest` annotation. This annotation will do a couple of things:

1. It will manage the [lifecycle](#engine-lifecycle) of the testcontainer / embedded test engine
2. It will create a client which can be used to interact with the engine.
3. It will (optionally) inject 3 fields in your test class:
   1. `ZeebeTestEngine` - This is the engine that will run your process. It will provide some basic functionality
      to help you write your tests, such as waiting for an idle state and increasing the time.
   2. `ZeebeClient` - This is the client that allows you to  send commands to the engine, such as

>       starting a process instance.

Reviewer: :wrench: Is this obvious or should we state it explicitly?
_The interface of this client is identical to the interface you use to connect to a real Zeebe engine._

   3. `RecordStream` - This gives you access to all the records that are processed by the engine.

>       It is what the assertions use for verifying expectations. This grants you the freedom to create your own assertions.

Reviewer: ```suggestion
      Assertions use the records to verify expectations. This grants you the freedom to create your own assertions.
```
(active voice)

Example:

```java

> // When using the embedded test engine

Reviewer: ```suggestion
// When using the embedded test engine (Java 17+)
```

import io.camunda.zeebe.process.test.extension.ZeebeProcessTest;

> // When using testcontainers

Reviewer: ```suggestion
// When using testcontainers (Java 8+)
```

We have defined an idle state as a state in which the engine makes no progress and is waiting for
new commands or events to trigger. Once the engine has detected it has become idle it will wait for
a delay (30ms) and check if it is still idle. If this is the case it is considered to be in idle
state and continue your test.

#### Wait for busy state

`engine.waitForBusyState(timeout)` will cause your test to stop executing until the engine has
reached a busy state. If the engine does not reach a busy state within the specified timeout a
`TimeoutException` will be thrown.

We consider the engine to have reached a busy state when any new record / command is processed since
we've started waiting.

Waiting for a busy state is useful in scenarios where you're expecting the engine to start doing
something, without explicitly triggering it yourself. An example of this would be a process with a
timer event. We can increase the time of the engine, but we cannot trigger the timer explicitly.
Because of this we should wait for a busy state after increasing the engine time.

## Examples

For example tests the best place to look right now is the tests in the QA module.

> ## Switching between testcontainers and embedded

Reviewer: :wrench: I think this should be sooner. Ideally right after the example where both imports are used.

Switching between testcontainers and embedded is very easy to do. You'll have to take two steps:

1. Switch to the relevant dependency
   - Testcontainers: `zeebe-process-test-extension-testcontainer`
   - Embedded: `zeebe-process-test-extension`
2. Change the import of `@ZeebeProcessTest`
   - Testcontainers: `import io.camunda.zeebe.process.test.extension.testcontainer.ZeebeProcessTest;`
   - Embedded: `import io.camunda.zeebe.process.test.extension.ZeebeProcessTest;`

> ## Engine lifecycle

Reviewer: :wrench: This could also appear as JavaDoc on the annotations, if it isn't already

The lifecycle of the engine will be fully managed by the extension. The lifecycle for both
extensions differs slightly.

**Testcontainers**

> 1. Before all tests start the testcontainer

Reviewer: The scope of all tests isn't fully clear. Is it a test class, a nested class, a test suite?

> 2. Before each test stop and recreate the Zeebe test engine

Reviewer: ```suggestion
2. Before each test stop the current Zeebe test engine (if applicable) and create a new Zeebe test engine for the next test
```

3. Engine
- This module contains the in memory engine. It is not bound to a specific Java version.
Therefore, it is not recommended to depend on this module.

> 4. Engine agent

Reviewer: :wrench: add links to each of the modules
